dan_layer/engine/src/crypto.rs

```diff
@@ -0,0 +1,103 @@
+//   Copyright 2023 The Tari Project
+//   SPDX-License-Identifier: BSD-3-Clause
+
+use std::borrow::Borrow;
+
+use lazy_static::lazy_static;
+use rand::rngs::OsRng;
+use tari_common_types::types::{
+    BulletRangeProof,
+    Commitment,
+    CommitmentFactory,
+    PrivateKey,
+    PublicKey,
+    RangeProofService,
+};
+use tari_crypto::{
+    commitment::{ExtensionDegree, HomomorphicCommitmentFactory},
+    extended_range_proof::ExtendedRangeProofService,
+    keys::{PublicKey as _, SecretKey},
+    ristretto::{
+        bulletproofs_plus::{RistrettoExtendedMask, RistrettoExtendedWitness},
+        RistrettoComSig,
+    },
+};
+use tari_template_lib::{crypto::RistrettoPublicKeyBytes, models::ConfidentialProof};
+use tari_utilities::ByteArray;
+
+lazy_static! {
+    static ref RANGE_PROOF_SERVICE: RangeProofService =
+        RangeProofService::init(64, 1, CommitmentFactory::default()).unwrap();
+}
+
+pub fn range_proof_service() -> &'static RangeProofService {
+    RANGE_PROOF_SERVICE.borrow()
+}
+
+pub fn generate_confidential_proof(mask: &PrivateKey, value: u64, minimum_value_promise: u64) -> ConfidentialProof {
+    let public_mask = PublicKey::from_secret_key(mask);
+    let (signature, commitment) = generate_commitment_knowledge_proof(mask, value);
+    let range_proof = generate_extended_bullet_proof(mask, value, minimum_value_promise);
```

Comment:
The range proof is a zero-knowledge proof about the commitment openings already, so maybe `knowledge_proof` must be optional? What is the use case for `knowledge_proof`?

Reply:
One aspect that may be worthwhile is that `public_mask` (actually just the public excess) is baked into the commitment signature, but I do not know if it has value. The validator nodes would need to verify the balance equation, and it would require the sum of the public excesses anyway as part of the Schnorr signature.

Reply:
For the mint use case, it makes sense, knowledge proof is not necessary. These are going to change somewhat to account for transfers which involve the balance proof.

Reply:
I've removed the proof of knowledge and the public mask/excess (which leaks the commitment value)